### ISSUE TITLE
Exported events are always longer than given duration (by 'time-before-event') and start with frozen time code

### ISSUE THREAD

**Issue by ocram** [Breakthrough/DVR-Scan#68]
When the text output lists an event like

```
|   Event #    |  Start Time  |   Duration   |   End Time   |
-------------------------------------------------------------
|  Event    1  |  00:15:46.8  |  00:00:24.2  |  00:16:11.1  |
```

(or the same time codes in the comma-separated list), the exported excerpts with motion are actually 29 seconds long (duration of 24 + `--time-before-event` of 5, probably). Is this expected?

Further, for the first 5 seconds (`--time-before-event`, probably), the time code in the upper left corner is frozen/static. Only after that, it starts to run (for the indicated duration of 24 seconds) until the end. But after just 3 seconds from the beginning, when the time code is still frozen for another 2 seconds, there is already motion in the video excerpt, while `--time-before-event` was supposed to be 5 seconds.

This happens not just for the very first event but for every event in any file. Thus, the output video is `numberOfEvents * 5` seconds longer than the sum of all printed durations.

Example of a command producing this behavior:

```
$ dvr-scan --input "my-input-video.mp4" --min-event-length "00:00:02.000" --time-before-event "00:00:05.000" --time-post-event "00:00:10.000" --time-code --output "my-output-video.avi"
```

DVR-Scan: v1.3 (latest release)
Python: v3.5.1
cv2: v4.2.0
numpy: v1.18.5
OS: Ubuntu 20.04
pip source: `https://bootstrap.pypa.io/pip/3.5/get-pip.py`
Video: h264 (High) (avc1 / 0x31637661), yuvj420p(pc), 1920x1080 [SAR 1:1 DAR 16:9], 990 kb/s, 30 fps, 30 tbr, 15360 tbn, 60 tbc (default)

**Comment by Breakthrough** [comment]
Indeed it looks like the timecode does have an issue with freezing, thank you for the report.

As for your first question, I'm not sure what you mean exactly - could you elaborate a bit or share a sample clip?  I'm not sure where the `numberOfEvents * 5` comes from given those numbers.

Thanks!

**Comment by ocram** [comment]
Thank you!

The first issue is the following: If you specify both `--time-before-event 5` and `--time-post-event 10` in combination with `--output <FILE>`, and the text output on the CLI lists one motion event like

```
|  Event    1  |  00:15:46.8  |  00:00:24.2  |  00:16:11.1  |
```

then what is the duration of your exported clip? For me, it is *not* 24.2 seconds, as listed, but 29 seconds. That means the CLI output does not match what’s exported. That is *probably* because `--time-post-event` is included in CLI output *and* for clip export whereas `--time-before-event` is considered only for clip export.

I know that only `--time-post-event` affects the actual motion detection (event continuation), but that doesn’t require the inconsistent and unintuitive behavior described above. The time after motion can easily affect motion detection (continuation window) while CLI output and exported clips are consistent. The docs agree that `--time-before-event` is supposed to have an effect on “the starting timecode of the current event when output”.

Since all this effects not just the very first event in a given input video but *any* event, the `--output <FILE>` clip’s length will exceed the sum of all durations given on the CLI by `numberOfEvents * 5`, probably `numberOfEvents * timeBeforeEvent`, repeating the effect described above for every single motion event in the input.

Further, the exported clip will *not* have `--time-before-event <TIME>` before any motion. As explained above, I specified 5 seconds as `--time-before-event` but after just 3 seconds there was motion included in the exported clip already. This does not depend on specific input files with special characteristics. It can be reproduced consistently for any input.

Incidentally, the 5 seconds, probably `--time-before-event`, is also exactly the duration during which the time code is frozen at the beginning of every event. That’s why I described all this in one issue, as it may be related.

**Comment by Breakthrough** [comment]
I see what you mean now, thanks for elaborating.  That is definitely not intentional, I was just confused at first since these are definitely two distinct bugs - but we can track them both in this issue.

Indeed, in your example the start of event 1 should probably be 00:10:46.8 with length 29.2 seconds.

> Further, the exported clip will not have --time-before-event <TIME> before any motion. As explained above, I specified 5 seconds as --time-before-event but after just 3 seconds there was motion included in the exported clip already. This does not depend on specific input files with special characteristics. It can be reproduced consistently for any input.

Now that seems like a separate bug/issue entirely, but I'll need to look more into that for sure.  Do you have any sample materials you can share?  And do you see the same issue if you both specify `-o` as well as if you omit `-o`?

**Comment by Breakthrough** [comment]
I think I also see another issue, where -tb should also join with a preceding scene.  This should all be much easier to deal with in DVR-Scan v2.0, but will need to figure out a solution for now for this for v1.4.

**Comment by ocram** [comment]
Thank you!

I think I tracked it down further:

 * More precisely, it seems that actual start times in exported clips (or duration of non-motion at the beginning) is `timeBeforeEvent - minEventLength`, e.g. with values such as `"00:00:02.000"`. I don’t think the latter being subtracted there makes any sense.
 * The CLI output (list of motion events, comma-separated time codes) does indeed consider `--time-post-event` while ignoring `--time-before-event`. So start times and durations in CLI output must include `--time-before-event` in their calculation.
 * Whether providing the `--output <FILE>` argument or omitting it, the list of events on the CLI is exactly the same. And, if the input video contains only a single motion event, the manually specified output file (argument) and the auto-generated output file (no argument) are byte-identical.
 * The time code in the upper left corner indeed seems to be frozen/static for the duration of `--time-before-event`.
 * As for `-tb` joining with a preceding scene, aren’t the docs explicit about the fact that `--time-post-event` affects the detection itself while `--time-before-event` does not?

**Comment by Breakthrough** [comment]
> More precisely, it seems that actual start times in exported clips (or duration of non-motion at the beginning) is timeBeforeEvent - minEventLength, e.g. with values such as "00:00:02.000". I don’t think the latter being subtracted there makes any sense.

Ah I see, yes that definitely is incorrect and not intended.

> As for -tb joining with a preceding scene, aren’t the docs explicit about the fact that --time-post-event affects the detection itself while --time-before-event does not?

Currently yes, but now I'm wondering if that makes sense.  Perhaps it should be left alone for now and configurable in the future, but I feel like it might be more intuitive if any motion events that overlap after applying -tb/-tp should be merged together.

Thanks for the comprehensive response, this was very helpful :)

**Comment by ocram** [comment]
> Currently yes, but now I'm wondering if that makes sense. Perhaps it should be left alone for now and configurable in the future, but I feel like it might be more intuitive if any motion events that overlap after applying -tb/-tp should be merged together.

I do not consider the current behavior intuitive or the best solution, but it’s well-documented and certainly something you can work with. I would agree that a conservative approach, as you described, is probably the best, i.e. not changing it for now and later starting with making it configurable.

**Comment by Breakthrough** [comment]
Thanks for the feedback - that's a huge help since I'm dealing with similar issues for a related project right now that DVR-Scan will be rewritten to use [1].

After giving it some thought and considering your feedback, I think in the next release the new default behavior *should* be to merge events if they overlap, instead of what happens now.  If you don't have any objections to that, I'm fine with changing it - I would much rather the default be a more intuitive approach until there is a configuration for it at least.

---------------------------

[1] With recent changes to PySceneDetect (specifically adding [a merge strategy](https://github.com/Breakthrough/PySceneDetect/blob/f1a28e10748fe27f59771082910f65cb35bf7017/scenedetect/scene_list.py#L47)), it should be possible to further specify another window where, after the events have been expanded, they can also be considered for merging.  For example, if there are two events detected at frames 10-20 and 50-60, and we set -tb=10 and -tp=15, the resulting events will be 0-35 and 40-75.  You could then set another parameter to merge any motion events that are N frames close, e.g. --merge=5, that way you can also set an effective distance between events for them to be considered the same (rather than just relying on them overlapping).  That way you can also ensure that for a given clip, you know there was no more motion detected for at least some time after the end of the event (otherwise they would be merged and the clip would have been longer).

**Comment by ocram** [comment]
A reasonable default behavior is always better than just more and more configuration options.

It’s just for backward compatibility – as someone might use this CLI application here extensively, rely on its exact behavior, and still want to upgrade to the latest versions – and because some bugs may have higher priority than changing behavior that works correctly (for now) that I thought a conservative approach *might* be good here.

> For example, if there are two events detected at frames 10-20 and 50-60, and we set -tb=10 and -tp=15, the resulting events will be 0-35 and 40-75.

That means both windows add up. So if `tb` and `tp` were both `15`, in this case, the two events would be merged, although they are `30` apart. That *could* be unintuitive as well, in my opinion.

The intuitive behavior I had in mind was that if *either* `tb` *or* `tp` makes an event extend to another event, they will be merged:

```
# (event_0_end + tp) >= event_1_start

or

(event_1_start - tb) <= event_0_end
or
(event_1_end + tp) >= event_2_start

or

# (event_2_start - tb) <= event_1_end
```

**Comment by Breakthrough** [comment]
That makes sense - I made the following visualization to add to the documentation for this proposed behavior:

![fig](https://user-images.githubusercontent.com/125316/146627591-aad14a43-823a-4f62-8e13-8478f21dbc51.png)

Does that align with your thoughts?  This also helps with PySceneDetect as I was aiming for something more complicated, but this seems like a much better idea. I had not considered this so thank you for bringing it up 👍

**Comment by ocram** [comment]
Thanks. This looks good, but I may need some clarification, to be honest:

Row 1 (tb=tp=0) and row 2 (tb=tp=5) are clearly two separate events, and row 4 (tb=tp=11) is clearly a single merged event, right?

But what exactly does row 3 (tb=tp=10) show? Do the two search distances overlap (thus a new color) but still not result in merging (thus not re-using one of the existing colors), which is the proposed new behavior?

One could also interpret this as a third event being created (which is not likely if you’re otherwise familiar with this application). But a new user could get that impression, since a colored block is an event in this diagram.

Perhaps, instead of overlapping (and color mixing), the bar can be divided:

![aa](https://user-images.githubusercontent.com/1681478/146631971-5e716857-081a-44f8-99d9-914a0fdd793e.png)

And then, maybe you’d even like to de-emphasize the search distances where the space is not really part of any event:

![aaa](https://user-images.githubusercontent.com/1681478/146631991-5650ac0c-baca-466e-ba36-7e16ec689bae.png)

Just an idea, though.

**Comment by Breakthrough** [comment]
That's a good idea, I managed to get it working with matplotlib.  Couldn't get the outlines to work properly, but this should be sufficient enough:

![fig](https://user-images.githubusercontent.com/125316/146651743-6bf66bf6-8165-44c0-b9fa-30d3cf3178cf.png)

![deemphasized](https://user-images.githubusercontent.com/125316/146651769-6b5a1b24-480e-4007-966d-30ec5d7d3236.png)

Now at least there's some code to auto-generate the diagrams, so thanks for the suggestion.  If the first row was omitted I think the de-emphasis would definitely help.  For now though I'm leaning towards the first version since it's a bit simpler and slightly more readability (especially for colourblind users).  There might be a way to change the doc stylesheet so that we could show the de-emphasized version on mouse hovering, but I'll defer that to later down the road.

**Comment by Breakthrough** [comment]
Thanks so much for your help with this.  I still need to figure out planning on when this will be possible, since there's a lot of technical debt to be addressed still which makes fixing this tricky.  However I should have some time during the holidays to come up with a plan, and will spend some time creating some test cases to demonstrate the results on actual videos.

Would love your opinion and feedback on some of the parts of the Python API I'm refactoring - please reach out to me on Discord if you're comfortable with that  (https://discord.gg/H83HbJngk7 - I still need to make a DVR-Scan specific chatroom so haven't linked to this on the frontpage yet).  And if not no worries - again I appreciate your help and feedback!

**Comment by Breakthrough** [comment]
Alright I've managed to fix most of these issues besides changing the behaviour of event merging.  I'll create a new issue to track that work and will likely be targeted for v2.0.

For v1.4, now the event start times, both in the CLI output as well as the video frames, will properly reflect both the event length `-l` and time before `-tb`.  Note that in the case that `-tp` is less than `-tb`, it may still be truncated if it overlaps with another motion event (since it does not affect merging).

**Comment by Breakthrough** [comment]
Are you sure that's correct?  I just did a test locally, here is the result with `-tb 1s`:

```[DVR-Scan] List of motion events:
-------------------------------------------------------------
|   Event #    |  Start Time  |   Duration   |   End Time   |
-------------------------------------------------------------
|  Event    1  |  00:00:01.2  |  00:00:07.9  |  00:00:09.1  |
|  Event    2  |  00:00:15.7  |  00:00:06.8  |  00:00:22.4  |
|  Event    3  |  00:00:32.3  |  00:00:04.6  |  00:00:36.9  |
-------------------------------------------------------------
[DVR-Scan] Comma-separated timecode values:
00:00:01.185,00:00:09.076,00:00:15.666,00:00:22.406,00:00:32.332,00:00:36.870```

And with `-tb 4s`:

```[DVR-Scan] List of motion events:
-------------------------------------------------------------
|   Event #    |  Start Time  |   Duration   |   End Time   |
-------------------------------------------------------------
|  Event    1  |  00:00:00.0  |  00:00:09.1  |  00:00:09.1  |
|  Event    2  |  00:00:12.7  |  00:00:09.8  |  00:00:22.4  |
|  Event    3  |  00:00:29.3  |  00:00:07.6  |  00:00:36.9  |
-------------------------------------------------------------
[DVR-Scan] Comma-separated timecode values:
00:00:00.000,00:00:09.076,00:00:12.663,00:00:22.406,00:00:29.329,00:00:36.870```

For event 3 above, we see the start time and duration have been shifted accordingly.  Are you seeing different results locally?  If so, can you please share the exact output and command line parameters you're testing with?

**Comment by ocram** [comment]
Pretty sure, yes, as I was specifically testing this multiple times due to the inconsistency.

I was using `v1.4` from pip. Any chance you have been using a slightly different code base, perhaps `master`?

As far as I remember, I was using

`--min-event-length "00:00:02.000" --time-before-event "00:00:03.000" --time-post-event "00:00:06.000"`

but might also have included `--frame-skip 2`.

**Comment by Breakthrough** [comment]
Ah sorry, it seems as though the issue is with frame skip only.  When calculating the event's start times, it is not taken into account.  I also see that there's a possibility that the timestamps can overlap in the case that -tb is too large, so will fix that up for v1.5 as well.  Will file an issue for that, thanks!

**Comment by Breakthrough** [comment]
Fixed in 0dab20b, see https://github.com/Breakthrough/DVR-Scan/issues/70#issuecomment-1037523434 for details.
